File: Android/app/src/main/java/com/pawsome/app/net/Firestore.kt
package com.example.pawsome.net

import com.example.pawsome.model.AppUser
import com.example.pawsome.model.Post
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.Query
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.tasks.await
import kotlinx.coroutines.withContext
import java.time.Instant

class FirestoreException(message: String) : Exception(message)

class Firestore {

    private val db = FirebaseFirestore.getInstance()

    suspend fun getPosts(limit: Int = 50): List<Post> = withContext(Dispatchers.IO) {
        db.collection("posts")
            .orderBy("timestamp", Query.Direction.DESCENDING)
            .limit(limit.toLong())
            .get()
            .await()
            .documents
            .mapNotNull { Post.fromDocument(it) }
    }

    suspend fun createPost(fields: Map<String, Any?>): String = withContext(Dispatchers.IO) {
        val ref = db.collection("posts").document()
        ref.set(fieldsWithFirestoreValues(fields)).await()
        ref.id
    }

    suspend fun deletePost(id: String) = withContext(Dispatchers.IO) {
        db.collection("posts").document(id).delete().await()
    }

    suspend fun toggleLike(postId: String, uid: String, like: Boolean) = withContext(Dispatchers.IO) {
        val ref = db.collection("posts").document(postId)
        ref.update("likes", if (like) FieldValue.arrayUnion(uid) else FieldValue.arrayRemove(uid)).await()
    }

    suspend fun getUser(uid: String): AppUser? = withContext(Dispatchers.IO) {
        val snapshot = db.collection("users").document(uid).get().await()
        if (!snapshot.exists()) null else AppUser.fromDocument(snapshot)
    }

    suspend fun updateUser(uid: String, fields: Map<String, Any?>) = withContext(Dispatchers.IO) {
        db.collection("users").document(uid).set(fieldsWithFirestoreValues(fields), com.google.firebase.firestore.SetOptions.merge()).await()
    }

    suspend fun fetchOrCreateUser(uid: String, name: String?, image: String?): AppUser =
        withContext(Dispatchers.IO) {
            getUser(uid)?.let { return@withContext it }

            val username = name ?: "User"
            val fields = mapOf(
                "username" to username,
                "profilePic" to (image ?: ""),
                "userNumber" to 0L,
                "createdAt" to FieldValue.serverTimestamp(),
            )
            db.collection("users").document(uid).set(fields).await()
            AppUser(uid, username, image, 0)
        }

    private fun fieldsWithFirestoreValues(fields: Map<String, Any?>): Map<String, Any?> =
        fields.mapValues { (_, value) ->
            when (value) {
                is Instant -> com.google.firebase.Timestamp(value.epochSecond, value.nano)
                else -> value
            }
        }

    private fun Any?.asFirestoreMap(): Any? = when (this) {
        is Instant -> com.google.firebase.Timestamp(this.epochSecond, this.nano)
        is List<*> -> map { it.asFirestoreMap() }
        is Map<*, *> -> entries.associate { it.key.toString() to it.value.asFirestoreMap() }
        else -> this
    }

    private fun fieldsWithFirestoreValuesCompat(fields: Map<String, Any?>): Map<String, Any?> =
        fields.mapValues { (_, value) -> value.asFirestoreMap() }
}

File: Firebase/iOS/Firestore.swift
import FirebaseFirestore

/// Apple-native Firestore adapter used by the iOS and macOS Pawsome targets.
/// Firestore schema follows the Android implementation exactly.
public final class PawsomeFirestore {
    public static let shared = PawsomeFirestore()
    private let db = Firestore.firestore()
    private init() {}

    public func getUser(uid: String) async throws -> [String: Any]? {
        let snapshot = try await db.collection("users").document(uid).getDocument()
        return snapshot.exists ? snapshot.data() : nil
    }

    public func createOrUpdateUser(uid: String, username: String?, profilePic: String?, loginMethod: String = "Unknown") async throws {
        let userRef = db.collection("users").document(uid)
        let counterRef = db.collection("counter").document("users")
        try await db.runTransaction { transaction, errorPointer in
            do {
                let existing = try transaction.getDocument(userRef)
                if existing.exists { return nil }
                let counterSnapshot = try transaction.getDocument(counterRef)
                let lastUserID = counterSnapshot.data()?["lastUserID"] as? Int ?? 0
                let nextUserID = lastUserID + 1
                transaction.setData(["lastUserID": nextUserID], forDocument: counterRef, merge: true)
                transaction.setData(["Username": username ?? "User", "ProfilePic": profilePic ?? "", "UserID": nextUserID, "LoginMethod": loginMethod, "JoinedOn": FieldValue.serverTimestamp()], forDocument: userRef, merge: false)
            } catch { errorPointer?.pointee = error as NSError }
            return nil
        }
    }

    public func updateUser(uid: String, fields: [String: Any]) async throws { try await db.collection("users").document(uid).setData(fields, merge: true) }

    public func getPosts(limit: Int = 50) async throws -> [[String: Any]] {
        let snapshot = try await db.collection("posts").order(by: "PostedAt", descending: true).limit(to: limit).getDocuments()
        return snapshot.documents.map { document in var data = document.data(); data["id"] = document.documentID; return data }
    }

    public func createPost(fields: [String: Any]) async throws -> String {
        var fields = fields
        guard fields["UserID"] != nil else { throw NSError(domain: "PawsomeFirestore", code: 1, userInfo: [NSLocalizedDescriptionKey: "UserID is required when creating a post"]) }
        if fields["PostedAt"] == nil { fields["PostedAt"] = FieldValue.serverTimestamp() }
        return try await db.collection("posts").addDocument(data: fields).documentID
    }

    public func deletePost(id: String) async throws { try await db.collection("posts").document(id).delete() }

    public func toggleLike(postId: String, uid: String, like: Bool) async throws {
        let value: Any = like ? FieldValue.arrayUnion([uid]) : FieldValue.arrayRemove([uid])
        try await db.collection("posts").document(postId).updateData(["likes": value])
    }
}
